File: 2025/20251009/소스2.py
# Initial Condition: left = 0, right = length - 1
# Termination: left == right
# Searching Left: right = mid
# Searching Right: left = mid+1


from typing import List

class Solution:
    def binarySearch(self, nums, left, right):
        n = len(nums)
        mid = (left + right) // 2

        if mid == 0:
            if nums[mid] > nums[mid + 1]:
                return mid
            else:
                if left == right:
                    return
                left = mid + 1
                return self.binarySearch(nums, left, right)

        elif mid == n - 1:
            if nums[mid] > nums[mid - 1]:
                return mid
            else:
                if left == right:
                    return
                right = mid
                return self.binarySearch(nums, left, right)

        else:
            if nums[mid] > nums[mid - 1] and nums[mid] > nums[mid + 1]:
                return mid

            else:
                if left == right:
                    return
                
                if nums[mid-1] < nums[mid+1]:
                    res1 = self.binarySearch(nums, mid + 1, right)
                    if res1 is not None:
                        return res1
                    return self.binarySearch(nums, left, mid)
                
                else:
                    res1 = self.binarySearch(nums, left, mid)
                    if res1 is not None:
                        return res1
                    return self.binarySearch(nums, mid + 1, right)
                    

    def findPeakElement(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return 0

        if len(nums) == 2:
            return nums.index(max(nums))

        n = len(nums)
        left = 0
        right = n - 1
        return self.binarySearch(nums, left, right)

File: 2025/20251101/소스2.js
//https://leetcode.com/problems/delete-nodes-from-linked-list-present-in-array/?envType=daily-question&envId=2025-11-01
//3217. Delete Nodes From Linked List Present in Array

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {number[]} nums
 * @param {ListNode} head
 * @return {ListNode}
 */
var binarySearch = (nums, target) => { 
    var left = 0
    var right = nums.length-1
    while (left <= right) { 
        const mid = Math.floor((left + right) / 2)
        if (nums[mid] == target) { 
            return mid
        }
        if (nums[mid] > target) { 
            right = mid-1
        } else { 
            left = mid+1
        }
    }
    return left
}

var modifiedList = function(nums, head) {
    var prev = null
    var curr = head
    var sortedNums = [...nums].sort((a, b) => a - b)

    while (curr != null)  {
        index = binarySearch(sortedNums, curr.val)
        if (index < sortedNums.length && sortedNums[index] == curr.val) { // 존재함
            if (prev) {
                prev.next = curr.next    
            } else {
                head = curr.next
            }
        } else { 
            prev = curr
        }
        curr = curr.next
    }
    return head
};